File: examples/python/examples.py
# ============================================
# Copilot Python Examples
# ============================================

# Example 1: Basic Function with Docstring
# ============================================

def is_palindrome(text):
    """
    Kiểm tra xem chuỗi có phải palindrome không
    
    Args:
        text (str): Chuỗi cần kiểm tra
        
    Returns:
        bool: True nếu là palindrome, False nếu không
        
    Example:
        >>> is_palindrome("A man, a plan, a canal: Panama")
        True
    """
    cleaned = text.lower()
    cleaned = ''.join(c for c in cleaned if c.isalnum())
    return cleaned == cleaned[::-1]


# Test
print(is_palindrome("A man, a plan, a canal: Panama"))  # True
print(is_palindrome("hello"))  # False


---

# Example 2: List Comprehension with Comment
# ============================================

# Lọc danh sách users: chỉ lấy admin
users = [
    {"id": 1, "name": "Alice", "role": "admin"},
    {"id": 2, "name": "Bob", "role": "user"},
    {"id": 3, "name": "Charlie", "role": "admin"}
]

# Filter role = 'admin'
# Copilot sẽ gợi ý list comprehension
admins = [user for user in users if user["role"] == "admin"]
print(admins)


---

# Example 3: Dictionary Operations
# ============================================

def select_fields(obj, fields):
    """
    Tạo dict mới với chỉ những field chỉ định
    
    Args:
        obj (dict): Object gốc
        fields (list): Danh sách field names
        
    Returns:
        dict: Dict mới với chỉ selected fields
        
    Example:
        >>> user = {"id": 1, "name": "Alice", "email": "alice@test.com"}
        >>> select_fields(user, ["name", "email"])
        {"name": "Alice", "email": "alice@test.com"}
    """
    return {field: obj[field] for field in fields if field in obj}


# Test
user = {"id": 1, "name": "Alice", "email": "alice@test.com", "password": "***"}
selected = select_fields(user, ["name", "email"])
print(selected)  # {"name": "Alice", "email": "alice@test.com"}


---

# Example 4: String Manipulation
# ============================================

def parse_name(full_name):
    """
    Chia full name thành first name, last name
    
    Args:
        full_name (str): Full name (ví dụ: "John Doe")
        
    Returns:
        dict: {"first_name": str, "last_name": str}
    """
    parts = full_name.strip().split()
    return {
        "first_name": parts[0] if parts else "",
        "last_name": " ".join(parts[1:]) if len(parts) > 1 else ""
    }


print(parse_name("John Doe"))  # {"first_name": "John", "last_name": "Doe"}
print(parse_name("Mary Jane Watson"))  # {"first_name": "Mary", "last_name": "Jane Watson"}


---

# Example 5: Async Function with Error Handling
# ============================================

import requests
import asyncio
from typing import Dict

async def fetch_user(user_id: int) -> Dict:
    """
    Fetch user từ API, xử lý lỗi
    
    Args:
        user_id (int): User ID
        
    Returns:
        dict: User data
        
    Raises:
        Exception: Nếu fetch fail
    """
    try:
        response = requests.get(f"https://api.example.com/users/{user_id}")
        response.raise_for_status()
        return response.json()
    except requests.RequestException as error:
        print(f"Error fetching user: {error}")
        raise


# Usage
# asyncio.run(fetch_user(1))


---

# Example 6: Email Validation
# ============================================

import re

def is_valid_email(email: str) -> bool:
    """
    Kiểm tra email có hợp lệ không
    
    Pattern: xxx@xxx.xxx
    
    Args:
        email (str): Email string
        
    Returns:
        bool: True nếu valid, False nếu invalid
    """
    pattern = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"
    return re.match(pattern, email) is not None


print(is_valid_email("test@example.com"))  # True
print(is_valid_email("invalid.email"))  # False


---

# Example 7: Class with Methods
# ============================================

class User:
    """User class with validation methods"""
    
    def __init__(self, user_id: int, name: str, email: str):
        """
        Initialize User
        
        Args:
            user_id (int): User ID
            name (str): User name
            email (str): User email
        """
        self.id = user_id
        self.name = name
        self.email = email
    
    def validate_email(self) -> bool:
        """Check if email is valid"""
        pattern = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"
        return re.match(pattern, self.email) is not None
    
    def to_dict(self) -> dict:
        """Return user as dictionary"""
        return {
            "id": self.id,
            "name": self.name,
            "email": self.email
        }
    
    def __repr__(self) -> str:
        return f"User(id={self.id}, name='{self.name}', email='{self.email}')"


# Usage
user = User(1, "Alice", "alice@test.com")
print(user.validate_email())  # True
print(user.to_dict())  # {"id": 1, "name": "Alice", "email": "..."}
print(user)  # User(id=1, name='Alice', email='alice@test.com')


---

# Example 8: List Operations with Functions
# ============================================

def get_numbers_in_range(numbers: list, min_val: int, max_val: int) -> list:
    """
    Lọc mảng numbers, chỉ lấy những số trong range [min_val, max_val]
    
    Args:
        numbers (list): Danh sách số
        min_val (int): Min value (inclusive)
        max_val (int): Max value (inclusive)
        
    Returns:
        list: Filtered numbers
    """
    return [n for n in numbers if min_val <= n <= max_val]


def get_unique_values(items: list) -> list:
    """
    Lấy unique values từ list (loại bỏ duplicates)
    
    Returns:
        list: Unique values
    """
    return list(set(items))


# Test
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(get_numbers_in_range(numbers, 3, 7))  # [3, 4, 5, 6, 7]
print(get_unique_values([1, 2, 2, 3, 3, 3]))  # [1, 2, 3]


---

# Example 9: Data Processing Pipeline
# ============================================

from typing import List

class DataProcessor:
    """Process data with multiple transformations"""
    
    @staticmethod
    def filter_active_users(users: List[Dict]) -> List[Dict]:
        """Lọc users có status = active"""
        return [u for u in users if u.get("status") == "active"]
    
    @staticmethod
    def map_to_usernames(users: List[Dict]) -> List[str]:
        """Extract usernames từ users"""
        return [u["name"] for u in users]
    
    @staticmethod
    def sort_by_name(users: List[Dict]) -> List[Dict]:
        """Sort users by name alphabetically"""
        return sorted(users, key=lambda u: u["name"])


# Usage
users = [
    {"id": 1, "name": "Alice", "status": "active"},
    {"id": 2, "name": "Bob", "status": "inactive"},
    {"id": 3, "name": "Charlie", "status": "active"}
]

active_users = DataProcessor.filter_active_users(users)
print(active_users)
# [{"id": 1, "name": "Alice", ...}, {"id": 3, "name": "Charlie", ...}]


---

# Example 10: Retry Decorator
# ============================================

import functools
import time

def retry(max_attempts=3, delay=1):
    """
    Decorator: retry function N lần nếu fail
    
    Args:
        max_attempts (int): Max retry times
        delay (int): Delay seconds giữa retries
    """
    def decorator(func):
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            attempts = 0
            while attempts < max_attempts:
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    attempts += 1
                    if attempts >= max_attempts:
                        raise
                    print(f"Retry {attempts}/{max_attempts-1}, waiting {delay}s...")
                    time.sleep(delay)
        return wrapper
    return decorator


@retry(max_attempts=3, delay=1)
def unstable_function():
    """Hàm có thể fail, retry 3 lần"""
    import random
    if random.random() < 0.7:
        raise Exception("Random failure!")
    return "Success!"


---

# Example 11: Unit Tests with pytest
# ============================================

import pytest

class Calculator:
    """Simple calculator"""
    
    @staticmethod
    def add(a, b):
        """Cộng hai số"""
        return a + b
    
    @staticmethod
    def multiply(a, b):
        """Nhân hai số"""
        return a * b
    
    @staticmethod
    def divide(a, b):
        """Chia hai số, raise error nếu b = 0"""
        if b == 0:
            raise ValueError("Cannot divide by zero")
        return a / b


# Test cases
def test_add():
    """Test add function"""
    assert Calculator.add(2, 3) == 5
    assert Calculator.add(-1, 1) == 0
    assert Calculator.add(0, 0) == 0


def test_multiply():
    """Test multiply function"""
    assert Calculator.multiply(3, 4) == 12
    assert Calculator.multiply(-2, 5) == -10


def test_divide():
    """Test divide function"""
    assert Calculator.divide(10, 2) == 5
    
    with pytest.raises(ValueError):
        Calculator.divide(10, 0)


---

# Example 12: Context Manager Pattern
# ============================================

class Database:
    """Simple database connection manager"""
    
    def __init__(self, connection_string):
        self.connection_string = connection_string
        self.conn = None
    
    def __enter__(self):
        """Connect when entering context"""
        print(f"Connecting to {self.connection_string}")
        self.conn = f"Connected to {self.connection_string}"
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Close connection when exiting context"""
        print("Closing connection")
        self.conn = None
    
    def query(self, sql):
        """Execute query"""
        if not self.conn:
            raise RuntimeError("Not connected")
        return f"Result of: {sql}"


# Usage
# with Database("postgresql://localhost/mydb") as db:
#     result = db.query("SELECT * FROM users")
#     print(result)
# # Automatically closes connection


---

# Example 13: Environment Variables & Config
# ============================================

import os
from typing import Optional

class Config:
    """Application configuration"""
    
    # Database
    DB_HOST = os.getenv("DB_HOST", "localhost")
    DB_PORT = int(os.getenv("DB_PORT", 5432))
    DB_NAME = os.getenv("DB_NAME", "mydb")
    
    # API
    API_URL = os.getenv("API_URL", "https://api.example.com")
    API_KEY = os.getenv("API_KEY")
    
    # App settings
    DEBUG = os.getenv("DEBUG", "False").lower() == "true"
    SECRET_KEY = os.getenv("SECRET_KEY")
    
    @classmethod
    def validate(cls) -> bool:
        """Validate required config"""
        required = ["API_KEY", "SECRET_KEY"]
        missing = [k for k in required if not os.getenv(k)]
        
        if missing:
            raise ValueError(f"Missing config: {missing}")
        return True


# Usage
# Config.validate()
# print(Config.DB_HOST)  # localhost


---

# Example 14: API Service Class
# ============================================

import requests
from typing import List, Dict, Optional

class UserAPIService:
    """Service to interact with User API"""
    
    def __init__(self, base_url: str, api_key: str):
        self.base_url = base_url
        self.api_key = api_key
        self.headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }
    
    def get_users(self) -> List[Dict]:
        """Get all users"""
        response = requests.get(
            f"{self.base_url}/users",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def get_user(self, user_id: int) -> Dict:
        """Get user by ID"""
        response = requests.get(
            f"{self.base_url}/users/{user_id}",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def create_user(self, user_data: Dict) -> Dict:
        """Create new user"""
        response = requests.post(
            f"{self.base_url}/users",
            json=user_data,
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def update_user(self, user_id: int, user_data: Dict) -> Dict:
        """Update user"""
        response = requests.put(
            f"{self.base_url}/users/{user_id}",
            json=user_data,
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def delete_user(self, user_id: int) -> bool:
        """Delete user"""
        response = requests.delete(
            f"{self.base_url}/users/{user_id}",
            headers=self.headers
        )
        response.raise_for_status()
        return True


# Usage
# service = UserAPIService("https://api.example.com", "your-api-key")
# users = service.get_users()
# new_user = service.create_user({"name": "John", "email": "john@test.com"})


---

# ============================================
# TIPS FOR COPILOT WITH PYTHON:
# ============================================
# 1. Dùng docstrings chi tiết (Args, Returns, Raises, Example)
# 2. Type hints (def func(name: str) -> bool:)
# 3. Comments mô tả ý tưởng trước code
# 4. PEP8 naming: snake_case for functions, UPPER_CASE for constants
# 5. Imports rõ ràng ở đầu file
# 6. Sử dụng List/Dict comprehensions
# 7. Context managers (__enter__, __exit__)
# ============================================

File: examples/javascript/examples.js
// ============================================
// Copilot JavaScript Examples
// ============================================

// Example 1: Basic Function with Copilot
// ============================================

// Hàm kiểm tra xem chuỗi có phải palindrome không
// Input: string - chuỗi cần kiểm tra
// Output: boolean - true nếu là palindrome, false nếu không
// Ví dụ: isPalindrome("racecar") => true
function isPalindrome(str) {
  const cleaned = str.toLowerCase().replace(/[^a-z0-9]/g, '');
  return cleaned === cleaned.split('').reverse().join('');
}

// Test
console.log(isPalindrome("A man, a plan, a canal: Panama")); // true
console.log(isPalindrome("hello")); // false

---

// Example 2: Array Methods with Comments
// ============================================

// Lọc mảng users, chỉ lấy những user có role = 'admin'
// Input: users (array), role (string)
// Output: filtered array
const users = [
  { id: 1, name: 'Alice', role: 'admin' },
  { id: 2, name: 'Bob', role: 'user' },
  { id: 3, name: 'Charlie', role: 'admin' }
];

function filterUsersByRole(users, role) {
  return users.filter(user => user.role === role);
}

const admins = filterUsersByRole(users, 'admin');
console.log(admins); // [{id:1, name:'Alice', role:'admin'}, {id:3, ...}]

---

// Example 3: Object Methods
// ============================================

// Tạo object mới với chỉ selected fields
// Input: user object, array of field names
// Output: new object với chỉ selected fields
// Ví dụ: selectFields(user, ['name', 'email']) => {name: '...', email: '...'}
function selectFields(obj, fields) {
  return fields.reduce((acc, field) => {
    acc[field] = obj[field];
    return acc;
  }, {});
}

const user = { id: 1, name: 'Alice', email: 'alice@test.com', password: '***' };
const selected = selectFields(user, ['name', 'email']);
console.log(selected); // {name: 'Alice', email: 'alice@test.com'}

---

// Example 4: String Manipulation
// ============================================

// Chia full name thành first name, last name
// Input: fullName (string) - ví dụ "John Doe"
// Output: object {firstName, lastName}
function parseName(fullName) {
  const parts = fullName.trim().split(/\s+/);
  return {
    firstName: parts[0],
    lastName: parts.slice(1).join(' ') || ''
  };
}

console.log(parseName("John Doe")); // {firstName: 'John', lastName: 'Doe'}
console.log(parseName("Mary Jane Watson")); // {firstName: 'Mary', lastName: 'Jane Watson'}

---

// Example 5: Async/Await with Error Handling
// ============================================

// Fetch user data từ API, xử lý lỗi
// Input: userId (number)
// Output: Promise<user object>
// Error: throw error nếu fetch fail
async function fetchUser(userId) {
  try {
    const response = await fetch(`https://api.example.com/users/${userId}`);
    if (!response.ok) {
      throw new Error(`HTTP Error: ${response.status}`);
    }
    const user = await response.json();
    return user;
  } catch (error) {
    console.error('Failed to fetch user:', error);
    throw error;
  }
}

// Usage
// fetchUser(1).then(user => console.log(user))
//   .catch(error => console.error(error));

---

// Example 6: Validation Function
// ============================================

// Kiểm tra email có hợp lệ không
// Pattern: xxx@xxx.xxx
// Return: boolean
function isValidEmail(email) {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

console.log(isValidEmail("test@example.com")); // true
console.log(isValidEmail("invalid.email")); // false
console.log(isValidEmail("test@domain")); // false

---

// Example 7: Class with Methods
// ============================================

class User {
  // Constructor: tạo user với id, name, email
  constructor(id, name, email) {
    this.id = id;
    this.name = name;
    this.email = email;
  }

  // Validate email format
  validateEmail() {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(this.email);
  }

  // Return object dạng JSON
  toJSON() {
    return {
      id: this.id,
      name: this.name,
      email: this.email
    };
  }
}

const user = new User(1, 'Alice', 'alice@test.com');
console.log(user.validateEmail()); // true
console.log(user.toJSON()); // {id: 1, name: 'Alice', email: '...'}

---

// Example 8: Map/Reduce/Filter Chain
// ============================================

const products = [
  { id: 1, name: 'Laptop', price: 1000, inStock: true },
  { id: 2, name: 'Mouse', price: 25, inStock: false },
  { id: 3, name: 'Keyboard', price: 75, inStock: true }
];

// Lấy sản phẩm còn hàng, tính giá giảm 10%, map thành {name, finalPrice}
const discountedProducts = products
  .filter(p => p.inStock)
  .map(p => ({ name: p.name, finalPrice: p.price * 0.9 }));

console.log(discountedProducts);
// [{name: 'Laptop', finalPrice: 900}, {name: 'Keyboard', finalPrice: 67.5}]

---

// Example 9: Debounce/Throttle Function
// ============================================

// Debounce: chạy function sau N ms khi user dừng gõ
// Input: function, delay (ms)
// Output: debounced function
function debounce(func, delay) {
  let timeoutId;
  return function (...args) {
    clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };
}

// Usage: search field
const handleSearch = debounce((query) => {
  console.log(`Searching for: ${query}`);
  // Call API
}, 500);

---

// Example 10: Promise with Error Handling
// ============================================

// Retry function N lần nếu fail
// Input: asyncFunc, maxRetries
// Output: Promise
function retryAsync(asyncFunc, maxRetries = 3) {
  return asyncFunc().catch(error => {
    if (maxRetries > 0) {
      console.log(`Retry... (${maxRetries} left)`);
      return retryAsync(asyncFunc, maxRetries - 1);
    }
    throw error;
  });
}

// Usage
// retryAsync(() => fetch('/api/data'), 3)
//   .then(res => console.log(res))
//   .catch(err => console.error(err));

---

// Example 11: Utility Functions
// ============================================

// Delay execution N milliseconds
// Return: Promise
function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// Deep clone object
function deepClone(obj) {
  return JSON.parse(JSON.stringify(obj));
}

// Check if object is empty
function isEmpty(obj) {
  return Object.keys(obj).length === 0;
}

// Get unique values from array
function getUnique(arr) {
  return [...new Set(arr)];
}

console.log(getUnique([1, 2, 2, 3, 3, 3])); // [1, 2, 3]

---

// Example 12: Event Emitter Pattern
// ============================================

class EventEmitter {
  constructor() {
    this.events = {};
  }

  // Subscribe to event
  on(event, callback) {
    if (!this.events[event]) {
      this.events[event] = [];
    }
    this.events[event].push(callback);
  }

  // Emit event
  emit(event, data) {
    if (this.events[event]) {
      this.events[event].forEach(callback => callback(data));
    }
  }

  // Unsubscribe from event
  off(event, callback) {
    if (this.events[event]) {
      this.events[event] = this.events[event].filter(cb => cb !== callback);
    }
  }
}

// Usage
const emitter = new EventEmitter();
emitter.on('message', (msg) => console.log(`Got: ${msg}`));
emitter.emit('message', 'Hello!'); // Got: Hello!

---

// Example 13: Unit Test Example (Jest)
// ============================================

describe('Calculator Functions', () => {
  // Hàm cộng hai số
  const add = (a, b) => a + b;

  // Test hàm add
  it('should add two numbers correctly', () => {
    expect(add(2, 3)).toBe(5);
    expect(add(-1, 1)).toBe(0);
    expect(add(0, 0)).toBe(0);
  });

  // Hàm nhân hai số
  const multiply = (a, b) => a * b;

  it('should multiply two numbers', () => {
    expect(multiply(3, 4)).toBe(12);
    expect(multiply(-2, 5)).toBe(-10);
  });
});

---

// Example 14: Module Export Pattern
// ============================================

// Các utility functions để export
const mathUtils = {
  // Tính tổng array
  sum: (arr) => arr.reduce((a, b) => a + b, 0),

  // Tính average
  average: (arr) => mathUtils.sum(arr) / arr.length,

  // Tìm max value
  max: (arr) => Math.max(...arr),

  // Tìm min value
  min: (arr) => Math.min(...arr)
};

// Export
// module.exports = mathUtils;

// Usage:
// const utils = require('./mathUtils');
// utils.sum([1, 2, 3]); // 6

---

// Example 15: Real-world API Service
// ============================================

class UserService {
  constructor(apiUrl) {
    this.apiUrl = apiUrl;
  }

  // Get all users
  async getUsers() {
    try {
      const response = await fetch(`${this.apiUrl}/users`);
      if (!response.ok) throw new Error('Failed to fetch');
      return await response.json();
    } catch (error) {
      console.error('Error fetching users:', error);
      throw error;
    }
  }

  // Get user by ID
  async getUserById(id) {
    try {
      const response = await fetch(`${this.apiUrl}/users/${id}`);
      if (!response.ok) throw new Error('User not found');
      return await response.json();
    } catch (error) {
      console.error(`Error fetching user ${id}:`, error);
      throw error;
    }
  }

  // Create new user
  async createUser(userData) {
    try {
      const response = await fetch(`${this.apiUrl}/users`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(userData)
      });
      if (!response.ok) throw new Error('Failed to create user');
      return await response.json();
    } catch (error) {
      console.error('Error creating user:', error);
      throw error;
    }
  }

  // Update user
  async updateUser(id, userData) {
    try {
      const response = await fetch(`${this.apiUrl}/users/${id}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(userData)
      });
      if (!response.ok) throw new Error('Failed to update user');
      return await response.json();
    } catch (error) {
      console.error(`Error updating user ${id}:`, error);
      throw error;
    }
  }

  // Delete user
  async deleteUser(id) {
    try {
      const response = await fetch(`${this.apiUrl}/users/${id}`, {
        method: 'DELETE'
      });
      if (!response.ok) throw new Error('Failed to delete user');
      return { success: true };
    } catch (error) {
      console.error(`Error deleting user ${id}:`, error);
      throw error;
    }
  }
}

// Usage
// const userService = new UserService('https://api.example.com');
// const users = await userService.getUsers();
// const newUser = await userService.createUser({name: 'John', email: 'john@test.com'});

---

// ============================================
// TIPS FOR COPILOT:
// ============================================
// 1. Viết comment chi tiết trước function
// 2. Dùng type hints (JSDoc comments)
// 3. Cho ví dụ cách sử dụng
// 4. Import/dependencies rõ ràng
// 5. Naming conventions consistent
// 6. Comments định nghĩa input/output/errors
// ============================================
